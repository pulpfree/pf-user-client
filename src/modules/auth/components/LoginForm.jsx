import React, { Component } from 'react'
import PropTypes from 'prop-types'

import g from 'glamorous'
import {
  Button,
} from 'react-md'
import { Field, SubmissionError } from 'redux-form'
import RenderTextField from '../../form/components/TextField'
import {
  email,
  password,
} from '../../form/utils'


class LoginForm extends Component {

  constructor(props) {
    super(props)
    this._onHandleSubmit = this._onHandleSubmit.bind(this)
  }

  _onHandleSubmit(values) {
    return this.props.onSave(values).then(response => {
      this.props.reset()
      return response
    }).catch(error => {
      throw new SubmissionError(error)
    })
  }

  render() {

    const {
      handleSubmit,
      pristine,
      submitting,
    } = this.props

    return (
      <g.Form
        onSubmit={handleSubmit(this._onHandleSubmit)}
        padding={'0 30px 30px'}
      >
        <Field
          autoFocus
          className="md-cell md-cell--12 md-cell--bottom form-field"
          component={RenderTextField}
          id="email"
          label="Email"
          name="email"
          placeholder="Enter email address"
          required
          type="email"
          validate={email}
        />
        <Field
          className="md-cell md-cell--12 md-cell--bottom form-field"
          component={RenderTextField}
          id="password"
          label="Password"
          name="password"
          placeholder="Enter your password"
          required
          style={{height: '90px'}}
          type="password"
          validate={password}
        />
        <Button
          className="md-cell md-cell--12 md-cell--bottom"
          disabled={pristine || submitting}
          id="submit"
          raised
          secondary
          style={{marginTop: '20px'}}
          type="submit"
        >
          {submitting ? 'Standby...' : 'Submit'}
        </Button>
      </g.Form>
    )
  }
}
LoginForm.propTypes = {
  handleSubmit: PropTypes.func.isRequired,
  onSave:       PropTypes.func.isRequired,
  pristine:     PropTypes.bool.isRequired,
  reset:        PropTypes.func.isRequired,
  submitting:   PropTypes.bool.isRequired,
}

export default LoginForm
