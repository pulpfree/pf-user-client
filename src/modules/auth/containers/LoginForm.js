// import React from 'react'
import { connect } from 'react-redux'
import { reduxForm } from 'redux-form'

import LoginFormComponent from '../components/LoginForm'


let LoginForm = reduxForm({ // eslint-disable-line no-class-assign
  form:           'login-form',
  // validate,
})(LoginFormComponent)

const mapStateToProps = null
const mapDispatchToProps = dispatch => {
  return {
    // This is just an example of an action that your submit handler might
    // fire. You can see I'm dispatching an action, just for example sake
    // to make it more realistic, but there's not even a reducer for the
    // action.
    onSave: contactFormValues => {
      console.log('Now running onSave action')
      return new Promise(resolve => {
        setTimeout(() => {
          console.log('Now dispatching action with type FORM_SUBMIT')
          dispatch({ type: 'FORM_SUBMIT', payload: contactFormValues })
          resolve()
        }, 1000)
      })
    },
  }
}

// NOTE: not clear if this is entirely necessary
// see: https://github.com/tylercollier/redux-form-test/blob/master/app/ContactFormContainer.js#L33 
// for explanation
const mergeProps = (stateProps, dispatchProps, ownProps) =>
  Object.assign({}, stateProps, dispatchProps, ownProps)

LoginForm = connect(
  mapStateToProps,
  mapDispatchToProps,
  mergeProps
)(LoginForm)

export default LoginForm