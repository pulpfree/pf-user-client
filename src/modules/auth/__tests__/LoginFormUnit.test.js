/* eslint no-undef: "off" */
import React from 'react'

import { shallow } from 'enzyme'
import { SubmissionError } from 'redux-form'

import LoginForm from '../components/LoginForm'

describe('LoginForm', () => {
  let mountedC
  let pristine, submitting, touched, error, reset, onSave, onSaveResponse, handleSubmit
  let wrapper

  const buildForm = () => {
    onSave = jest.fn().mockReturnValue(onSaveResponse)
    const props = {
      onSave,
      submitting,
      pristine,
      // The real redux form has many properties for each field,
      // including onChange and onBlur handlers. We only need to provide
      // the ones that will change the rendered output.
      fields: {
        email: {
          value: 'me@there.co',
          touched,
          error,
        },
      },
      handleSubmit,
      reset,
    }
    if (!mountedC) {
      mountedC = shallow(<LoginForm {...props} />)
    }
    return mountedC
  }

  beforeEach(() => {
    error = null
    handleSubmit = fn => fn
    mountedC = undefined
    onSaveResponse = Promise.resolve()
    pristine = false
    reset = jest.fn()
    submitting = false
    touched = false
  })

  it('always renders fields', () => {
    wrapper = buildForm()
    expect(wrapper.find('Field').length).toBe(2)
    expect(wrapper.find('Field[id="email"]').length).toEqual(1)
    expect(wrapper.find('Field[id="password"]').length).toEqual(1)
  })

  it('submits successfully', () => {
    wrapper = buildForm()
    wrapper.simulate('submit')

    return onSaveResponse.then(() => {
      expect(reset.mock.calls.length).toEqual(1)
    })
  })

  it('matches snapshot', () => {
    expect(buildForm()).toMatchSnapshot()
  })

  describe('Login form errors', () => {

    beforeEach(() => {
      onSaveResponse = Promise.reject('some rejection')
    })

    it('throws a SubmissionError on error in the form submit handler', () => {
      let promiseReturnedFromFormHandler
      handleSubmit = fn => {
        return function() {
          promiseReturnedFromFormHandler = fn(arguments)
        }
      }
      wrapper = buildForm()
      wrapper.simulate('submit')
      expect(onSave.mock.calls.length).toEqual(1)
      return promiseReturnedFromFormHandler.then(() => {
        throw new Error('Submission error should have been checked but was not')
      }).catch(error => {
        expect(error).toBeInstanceOf(SubmissionError)
      })
    })

  })

})
