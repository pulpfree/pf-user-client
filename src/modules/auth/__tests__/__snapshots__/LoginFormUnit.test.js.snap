// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LoginForm matches snapshot 1`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <LoginForm
    fields={
        Object {
            "email": Object {
              "error": null,
              "touched": false,
              "value": "me@there.co",
            },
          }
    }
    handleSubmit={[Function]}
    onSave={[Function]}
    pristine={false}
    reset={[Function]}
    submitting={false}
/>,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "class",
    "props": Object {
      "children": Array [
        <Field
          autoFocus={true}
          className="md-cell md-cell--12 md-cell--bottom form-field"
          component={[Function]}
          id="email"
          label="Email"
          name="email"
          placeholder="Enter email address"
          required={true}
          type="email"
          validate={[Function]}
/>,
        <Field
          className="md-cell md-cell--12 md-cell--bottom form-field"
          component={[Function]}
          id="password"
          label="Password"
          name="password"
          placeholder="Enter your password"
          required={true}
          style={
                    Object {
                              "height": "90px",
                            }
          }
          type="password"
          validate={[Function]}
/>,
        <withInk(withTooltip(Button))
          className="md-cell md-cell--12 md-cell--bottom"
          disabled={false}
          id="submit"
          inkTransitionEnterTimeout={450}
          inkTransitionLeaveTimeout={300}
          inkTransitionOverlap={150}
          raised={true}
          secondary={true}
          style={
                    Object {
                              "marginTop": "20px",
                            }
          }
          type="submit"
>
          Submit
</withInk(withTooltip(Button))>,
      ],
      "onSubmit": [Function],
      "padding": "0 30px 30px",
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "class",
        "props": Object {
          "autoFocus": true,
          "className": "md-cell md-cell--12 md-cell--bottom form-field",
          "component": [Function],
          "id": "email",
          "label": "Email",
          "name": "email",
          "placeholder": "Enter email address",
          "required": true,
          "type": "email",
          "validate": [Function],
        },
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "class",
        "props": Object {
          "className": "md-cell md-cell--12 md-cell--bottom form-field",
          "component": [Function],
          "id": "password",
          "label": "Password",
          "name": "password",
          "placeholder": "Enter your password",
          "required": true,
          "style": Object {
            "height": "90px",
          },
          "type": "password",
          "validate": [Function],
        },
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "class",
        "props": Object {
          "children": "Submit",
          "className": "md-cell md-cell--12 md-cell--bottom",
          "disabled": false,
          "id": "submit",
          "inkTransitionEnterTimeout": 450,
          "inkTransitionLeaveTimeout": 300,
          "inkTransitionOverlap": 150,
          "raised": true,
          "secondary": true,
          "style": Object {
            "marginTop": "20px",
          },
          "type": "submit",
        },
        "ref": null,
        "rendered": "Submit",
        "type": [Function],
      },
    ],
    "type": [Function],
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "class",
      "props": Object {
        "children": Array [
          <Field
            autoFocus={true}
            className="md-cell md-cell--12 md-cell--bottom form-field"
            component={[Function]}
            id="email"
            label="Email"
            name="email"
            placeholder="Enter email address"
            required={true}
            type="email"
            validate={[Function]}
/>,
          <Field
            className="md-cell md-cell--12 md-cell--bottom form-field"
            component={[Function]}
            id="password"
            label="Password"
            name="password"
            placeholder="Enter your password"
            required={true}
            style={
                        Object {
                                    "height": "90px",
                                  }
            }
            type="password"
            validate={[Function]}
/>,
          <withInk(withTooltip(Button))
            className="md-cell md-cell--12 md-cell--bottom"
            disabled={false}
            id="submit"
            inkTransitionEnterTimeout={450}
            inkTransitionLeaveTimeout={300}
            inkTransitionOverlap={150}
            raised={true}
            secondary={true}
            style={
                        Object {
                                    "marginTop": "20px",
                                  }
            }
            type="submit"
>
            Submit
</withInk(withTooltip(Button))>,
        ],
        "onSubmit": [Function],
        "padding": "0 30px 30px",
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "class",
          "props": Object {
            "autoFocus": true,
            "className": "md-cell md-cell--12 md-cell--bottom form-field",
            "component": [Function],
            "id": "email",
            "label": "Email",
            "name": "email",
            "placeholder": "Enter email address",
            "required": true,
            "type": "email",
            "validate": [Function],
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "class",
          "props": Object {
            "className": "md-cell md-cell--12 md-cell--bottom form-field",
            "component": [Function],
            "id": "password",
            "label": "Password",
            "name": "password",
            "placeholder": "Enter your password",
            "required": true,
            "style": Object {
              "height": "90px",
            },
            "type": "password",
            "validate": [Function],
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "class",
          "props": Object {
            "children": "Submit",
            "className": "md-cell md-cell--12 md-cell--bottom",
            "disabled": false,
            "id": "submit",
            "inkTransitionEnterTimeout": 450,
            "inkTransitionLeaveTimeout": 300,
            "inkTransitionOverlap": 150,
            "raised": true,
            "secondary": true,
            "style": Object {
              "marginTop": "20px",
            },
            "type": "submit",
          },
          "ref": null,
          "rendered": "Submit",
          "type": [Function],
        },
      ],
      "type": [Function],
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
      },
    },
  },
}
`;
